tests/integration/synapseclient/models/async/test_json_schema_async.py

```diff
@@ -0,0 +1,481 @@
+        finally:
+            # Clean up the JSON schema binding
+            await created_folder.delete_json_schema_from_entity_async(
+                synapse_client=self.syn
+            )
+
+    async def test_delete_json_schema_from_entity_async(
+        self,
+        folder: Folder,
+        project_model: Project,
+        create_test_organization_with_schema: Tuple[JsonSchemaOrganization, str],
+    ) -> None:
+        """Test deleting a bound JSON schema from an entity."""
+        # Create the folder
+        created_folder = await folder.store_async(parent=project_model)
+        self.schedule_for_cleanup(folder.id)
+
+        # Bind the JSON schema to the folder
+        _, test_product_schema_uri = create_test_organization_with_schema
+        await created_folder.bind_json_schema_to_entity_async(
+            json_schema_uri=test_product_schema_uri, synapse_client=self.syn
+        )
+
+        # Delete the JSON schema from the folder
+        await created_folder.delete_json_schema_from_entity_async(
+            synapse_client=self.syn
+        )
+
+        # Verify that the JSON schema is no longer bound
+        with pytest.raises(
+            SynapseHTTPError,
+            match=f"404 Client Error: No JSON schema found for '{folder.id}'",
+        ):
+            await created_folder.get_json_schema_from_entity_async(
+                synapse_client=self.syn
+            )
+
+    async def test_get_json_schema_derived_keys_async(
+        self,
+        folder: Folder,
+        project_model: Project,
+        create_test_organization_with_schema: Tuple[
+            JsonSchemaOrganization,
+            str,
+        ],
+    ) -> None:
+        """Test retrieving derived keys from a bound JSON schema."""
+        # Create the folder
+        created_folder = await folder.store_async(parent=project_model)
+        self.schedule_for_cleanup(folder.id)
+        _, test_product_schema_uri = create_test_organization_with_schema
+
+        # Bind the JSON schema to the folder
+        try:
+            await created_folder.bind_json_schema_to_entity_async(
+                json_schema_uri=test_product_schema_uri,
+                enable_derived_annos=True,
+                synapse_client=self.syn,
+            )
+
+            # Store annotations
+            created_folder.annotations = {
+                "productDescription": "test description here",
+                "productQuantity": 100,
+                "productPrice": 100,
+            }
+
+            stored_folder = await folder.store_async(parent=project_model)
+            response = await created_folder.get_json_schema_from_entity_async(
+                synapse_client=self.syn
+            )
+            response = await created_folder.get_json_schema_from_entity_async(
+                synapse_client=self.syn
+            )
+            assert response["enableDerivedAnnotations"] == True
+
+            sleep(2)
```

Comment:
I tried removing it and created another function that wraps around the async call to ensure that the async call finished. Even though the wrapper function worked as expected, the test would fail. This is something that needs to be further investigated.

Reply:
Instead of sleep (Which will sleep the thread this is running on), the preferred way is to use asyncio:

`await asyncio.sleep(2)`


It likely fails WITHOUT the sleep because of the server side process being eventually consistent (Meaning that clients have to wait until the server side state is updated with the changes you made). Even though the Job you submitted to Synapse completed, anything that job updated may not have been fully updated yet.

Using a wait statement here is the right call (And something we've have to do in other integration tests).

Reply:
Wow thanks for the explanation. I didn't realize that the job could be completed but the update might not be finished.